src/uct/ugni/udt/ugni_udt_iface.c

```diff
@@ -28,125 +30,138 @@ static ucs_config_field_t uct_ugni_udt_iface_config_table[] = {
-                                     new_desc, goto exit);
-            /* set the new desc */
-            iface->desc_any = new_desc;
-        }
-        status = uct_ugni_udt_ep_any_post(iface);
-        if (ucs_unlikely(UCS_OK != status)) {
-            ucs_error("Failed to post uct_ugni_udt_ep_any_post");
-            goto exit;
+    if(GNI_POST_TERMINATED == post_state) {
+        if(UCT_UGNI_UDT_ANY == id) {
+            return UCS_ERR_CANCELED;
+        } else {
+            return UCS_OK;
         }
-    } else {
-        /* Ack message */
-        ucs_assert_always(NULL != ep);
+    }
 
-        if (header->type == UCT_UGNI_UDT_PAYLOAD) {
-            /* data message was received */
-            uct_iface_trace_am(&iface->super.super, UCT_AM_TRACE_TYPE_RECV,
-                               header->am_id, payload, header->length, "RX: AM");
-            pthread_mutex_unlock(&uct_ugni_global_lock);
-            status = uct_iface_invoke_am(&iface->super.super, header->am_id, payload,
-                                         header->length, user_desc);
-            pthread_mutex_lock(&uct_ugni_global_lock);
-            if (UCS_OK == status) {
-                uct_ugni_udt_reset_desc(desc, iface);
-                ucs_mpool_put(desc);
-            } else {
-                /* set iface for a later release call */
-                uct_recv_desc_iface(user_desc) = &iface->super.super.super;
-            }
-        }
-        /* no data, just an ack */
+    if(GNI_POST_COMPLETED != post_state) {
+        ucs_error("GNI_EpPostDataWaitById gave unexpected response: %u", post_state);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if(UCT_UGNI_UDT_ANY != id) {
         --iface->super.outstanding;
         --ep->super.outstanding;
+        uct_ugni_ep_check_flush(&ep->super);
+    }
+
+    header = uct_ugni_udt_get_rheader(desc, iface);
+
+    ucs_trace("Got datagram id: %lu type: %i len: %i am_id: %i", id, header->type, header->length, header->am_id);
+
+    if(UCT_UGNI_UDT_PAYLOAD != header->type) {
+        /* ack message, no data */
+        ucs_assert_always(NULL != ep);
+        ucs_mpool_put(ep->posted_desc);
+        uct_ugni_ep_check_flush(&ep->super);
         ep->posted_desc = NULL;
+        return UCS_OK;
     }
 
-exit:
-    pthread_mutex_unlock(&uct_ugni_global_lock);
-    /* have a go a processing the pending queue */
-    ucs_arbiter_dispatch(&iface->super.arbiter, 1, uct_ugni_ep_process_pending, NULL);
+    return UCS_INPROGRESS;
+}
+
+static void *uct_ugni_udt_device_thread(void *arg)
+{
+    uct_ugni_udt_iface_t *iface = (uct_ugni_udt_iface_t *)arg;
+    gni_return_t ugni_rc;
+    uint64_t id;
+    int dummy;
+
+    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, &dummy);
+    while(1){
+        pthread_mutex_lock(&iface->device_lock);
+        while(iface->events_ready) {
+            ucs_trace("Waiting on condition");
+            pthread_cond_wait(&iface->device_condition, &iface->device_lock);
```

Comment:
why need pthread_cond_wait if you are waiting anyway with GNI_PostdataProbeWaitById()?